### PR TITLE
[API] As a logged in user, I can fetch my keyword lists

### LINKED ISSUE #15: [API] As a logged in user, I can fetch my keyword lists
## Why

Users able to fetch the his/her keyword list via the API

- This API should support pagination also
- The API route should be `GET /api/v1/keywords`
- This API should support query params also to support searching `/api/v1/keywords?search=your keyword`
 
## Who Benefits?

Users

### PR DESCRIPTION
**Pull request by junan**
https://github.com/junan/elixir_google_scraper/projects/1#card-61315817

## What happened

User can fetch keywords
 
## Insight

N/A
 
## Proof Of Work

Screen Shot 2021-07-08 at 1 16 21 AM

### With search param
Screen Shot 2021-07-08 at 1 16 00 AM

### CODE REVIEW

File: test/elixir_google_scraper_web/controllers/api/v1/keyword_controller_test.exs [hunk 1]
```diff
@@ -1,24 +1,81 @@
 defmodule ElixirGoogleScraperWeb.Api.V1.KeywordControllerTest do
   use ElixirGoogleScraperWeb.ConnCase, async: true
+  use ElixirGoogleScraperWeb.ApiCase, async: true
 
   alias ExOauth2Provider.AccessTokens
   alias ExOauth2Provider.Applications
 
+  describe "GET index/2" do
+    test "returrns 200 status with list of keywords", %{conn: conn} do
```
Comment (hanam1ni) [inline test/elixir_google_scraper_web/controllers/api/v1/keyword_controller_test.exs]:
```suggestion
    test "returns 200 status with list of keywords", %{conn: conn} do
```

File: test/elixir_google_scraper_web/controllers/api/v1/keyword_controller_test.exs [hunk 2]
```diff
@@ -1,24 +1,81 @@
...
+                   "relationships" => %{},
+                   "type" => "keyword"
+                 }
+               ],
+               "included" => [],
+               "meta" => %{"page" => _, "page_size" => _, "pages" => _, "records" => _}
+             } = json_response(conn, 200)
+    end
+
+    test "returrns only matched keywords when search param is present", %{conn: conn} do
```
Comment (hanam1ni) [inline test/elixir_google_scraper_web/controllers/api/v1/keyword_controller_test.exs]:
```suggestion
    test "returns only matched keywords when given search params", %{conn: conn} do
```

File: test/support/api_case.ex [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+      alias ExOauth2Provider.Applications
+
+      def authenticated_request(user, conn, target, type, params \\ %{}) do
+        attrs = %{name: "Application", redirect_uri: "https://example.org/endpoint"}
+        file = %Plug.Upload{content_type: "text/csv", path: "test/fixture/keywords.csv"}
+
+        {_, oauth_app} =
+          Applications.create_application(nil, attrs, otp_app: :elixir_google_scraper)
+
+        {_, acess_token} =
```
Comment (hanam1ni) [inline test/support/api_case.ex]:
```suggestion
        {_, access_token} =
```

File: test/support/api_case.ex [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  use ExUnit.CaseTemplate
+
+  using do
+    quote do
+      alias ExOauth2Provider.AccessTokens
+      alias ExOauth2Provider.Applications
+
+      def authenticated_request(user, conn, target, type, params \\ %{}) do
+        attrs = %{name: "Application", redirect_uri: "https://example.org/endpoint"}
+        file = %Plug.Upload{content_type: "text/csv", path: "test/fixture/keywords.csv"}
```
Comment (hanam1ni) [inline test/support/api_case.ex]: Are we using this one?

File: test/support/api_case.ex [hunk 5]
```diff
@@ -0,0 +1,33 @@
+defmodule ElixirGoogleScraperWeb.ApiCase do
+  use ExUnit.CaseTemplate
+
+  using do
+    quote do
+      alias ExOauth2Provider.AccessTokens
+      alias ExOauth2Provider.Applications
+
+      def authenticated_request(user, conn, target, type, params \\ %{}) do
```
Comment (hanam1ni) [inline test/support/api_case.ex]: How about use the `setup` block to return authenticated `conn` instead? Then in the test, we use that `conn` to make a request. 

You can check about setup: block here https://hexdocs.pm/phoenix/testing.html#the-conncase

Comment (junan) [reply test/support/api_case.ex]: As discussed, I used a helper function to authenticate the `conn`

File: lib/elixir_google_scraper_web/controllers/api/v1/keyword_controller.ex [hunk 6]
```diff
@@ -3,6 +3,13 @@ defmodule ElixirGoogleScraperWeb.Api.V1.KeywordController do
 
   alias ElixirGoogleScraper.Scraper.Keywords
   alias ElixirGoogleScraperWeb.V1.ErrorView
+  alias ElixirGoogleScraperWeb.V1.KeywordView
```
Comment (rosle) [inline lib/elixir_google_scraper_web/controllers/api/v1/keyword_controller.ex]:
```suggestion
  alias ElixirGoogleScraperWeb.V1.{ErrorView, KeywordView}
```

File: test/elixir_google_scraper_web/controllers/api/v1/keyword_controller_test.exs [hunk 7]
```diff
@@ -1,23 +1,85 @@
 defmodule ElixirGoogleScraperWeb.Api.V1.KeywordControllerTest do
   use ElixirGoogleScraperWeb.ConnCase, async: true
+  use ElixirGoogleScraperWeb.ApiCase, async: true
```
Comment (rosle) [inline test/elixir_google_scraper_web/controllers/api/v1/keyword_controller_test.exs]: Having these 2 looks a bit weird. I think we have 2 option here

1. Use `ApiCase` -- Remove `ConnCase` from here and add everything we need in `ApiCase`
2. Use `ConnCase ` -- Create another module in test support. Maybe we can call `AuthenticationHelper` and move all helpers there.

```elixir
def register_and_log_in_user ...
def login ...
def authenticated_api_conn ...
```

Then we include this helper to conn case

```elixir
import ElixirGoogleScraperWeb.AuthenticationHelper
```

**Review by hanam1ni — APPROVED** [review]

File: test/support/auth_helper_case.ex [hunk 8]
```diff
@@ -0,0 +1,50 @@
+defmodule ElixirGoogleScraperWeb.AuthHelperCase do
+  use ExUnit.CaseTemplate
```
Comment (rosle) [inline test/support/auth_helper_case.ex]:
```suggestion
defmodule ElixirGoogleScraperWeb.AuthHelper do
```
I think we can use a simple module here. It just a collection of helper functions. 👍

**Review by rosle — APPROVED** [review]

**Pull request merged**